include/sparrow/array_data_factory.hpp

```diff
@@ -0,0 +1,347 @@
+            [&set_index](const T& value)
+            {
+                return set_index[std::cref(value)];
+            }
+        );
+
+        std::transform(
+            set_index.begin(),
+            set_index.end(),
+            std::back_inserter(result.values),
+            [](const auto& pair)
+            {
+                return pair.first;
+            }
+        );
+
+        return result;
+    }
+
+    template <typename T>
+    array_data for_dictionary_encoded_layout()
+    {
+        return {
+            .type = data_descriptor(arrow_type_id<std::uint64_t>()),
+            .length = 0,
+            .offset = 0,
+            .bitmap = {},
+            .buffers = {array_data::buffer_type(sizeof(std::int64_t), 0)},
+            .child_data = {},
+            .dictionary = value_ptr<array_data>(for_variable_size_binary_layout<T>())
+        };
+    }
+
+    template <std::ranges::range ValueRange>
+    array_data
+    for_dictionary_encoded_layout(const ValueRange& values, const array_data::bitmap_type& bitmap, std::int64_t offset)
+    {
+        SPARROW_ASSERT_TRUE(values.size() == bitmap.size());
+        SPARROW_ASSERT_TRUE(std::cmp_greater_equal(values.size(), offset));
+
+        const auto vec_and_indexes = ranges_to_vec_and_indexes(values);
+        const auto& indexes = vec_and_indexes.indexes;
+        static const auto create_buffer = [&indexes]()
+        {
+            const size_t buffer_size = indexes.size() * sizeof(size_t) / sizeof(uint8_t);
+            array_data::buffer_type b(buffer_size);
+            std::ranges::copy(indexes, b.data<size_t>());
+            return b;
+        };
+        return {
+            .type = data_descriptor(arrow_type_id<std::uint64_t>()),
+            .length = static_cast<array_data::length_type>(indexes.size()),
+            .offset = offset,
+            .bitmap = bitmap,
+            .buffers = {create_buffer()},
+            .child_data = {},
+            .dictionary = value_ptr<array_data>(for_variable_size_binary_layout(
+                vec_and_indexes.values,
+                array_data::bitmap_type(vec_and_indexes.values.size(), true),
+                0
+            ))
+        };
+    }
+
+    template <class Layout>
+    array_data default_array_data_factory()
+    {
+        if constexpr (mpl::InstantiationOf<fixed_size_layout, Layout>)
+        {
+            return for_fixed_size_layout<typename Layout::inner_value_type>();
+        }
+        else if constexpr (mpl::InstantiationOf<variable_size_binary_layout, Layout>)
+        {
+            return for_variable_size_binary_layout<typename Layout::inner_value_type>();
+        }
+        else if constexpr (mpl::InstantiationOf<dictionary_encoded_layout, Layout>)
+        {
+            return for_dictionary_encoded_layout<typename Layout::inner_value_type>();
+        }
+        else
+        {
+            static_assert(!std::is_same_v<Layout, Layout>, "Unsupported layout type");
```

Comment:
Consider making a type trait which names clarify the intent of this kind of check pattern.